### ISSUE TITLE
Enhancement: (Capture rate driven) Playback Auto Rate Family support

### ISSUE THREAD

**Issue by yo3fxy** [marcoevang/camilladsp-setrate#2]
Hi,

I use CamillaDSP for upsampling to 705600 / 768000 and I love the ability to automatically change capture sample rate using your tool. Would it be possible to add support for changing both (Playback & Capture) sample rates, playback sample rate being an integer multiple of 44100 / 48000 depending on the capture sample rate?

Thanks.

**Comment by marcoevang** [comment]
Hello,
so you would like to set the capture sample rate according to the sample rate of the input signal, and then dynamically set the playback sample rate to an integer multiple of the capture sample rate. 
Did I get that right? If yes, can you describe a use case where this might be useful? In the next version of camilladsp-setrate I might consider adding this feature.

**Comment by yo3fxy** [comment]
yes, you got it right; i would like to automatically switch capture samplerate, as well as playback samplerate (in the same rate family as the capture rate).

My current chain is iPhone -> (USB Gadget) rPi4 with CamillaDSP (USB) -> Topping D50s, with automatic capture samplerate switching. I tried to use CamillaDSP with AccurateAsync resampling (which is supposed to have similar performance/accuracy as the Synchronous resampler), but my rPI4 is not powerfull enough and i had to "downgrade" to the BalancedAsync resampler. With auto playback rate-family switching i could use Synchrpnous resampler instead, which would be less taxing on the CPU (and my passively cooled rPI4).

**Comment by sirblew** [comment]
> If yes, can you describe a use case where this might be useful? In the next version of camilladsp-setrate I might consider adding this feature.

Is it not more efficient to use an integer multiple when upsampling anyway? Eg 2fs or 4fs? My understanding is it requires less CPU power to upsample as an integer.

**Comment by marcoevang** [comment]
Just released version v2.1.0  adds the ability to upsample the incoming audio by a fixed factor (e.g. 2X, 4X, ...), which is the enhancement requested by @yo3fxy.
@yo3fxy , please let me know if this works for you as expected.

**Comment by GitHubGeek** [comment]
Is it possible to make the factor dynamic, similar to how Squeezelite resample work?

E.g. DAC can accept up to 768KHz. 

- 44.1KHz capture -> 705.6KHz output (16x)
- 48KHz -> 768KHz (16x)
- 88.2KHz -> 705.6KHz (8x)
- 96KHz -> 768KHz (8x)
- 192KHz -> 768KHz (4x)

**Comment by marcoevang** [comment]
If I understand well, you mean oversampling to the highest sampling rate the DAC can accept, regardless of the source sampling frequency.
No, camilladsp-setrate does not sport this feature. However, if I find some spare time, I'll try to implement it.
I'll inform of this on github and on the dedicated thread on diyaudio.com.

**Comment by GitHubGeek** [comment]
> If I understand well, you mean oversampling to the highest sampling rate the DAC can accept, regardless of the source sampling frequency.

Yes this is the goal, while maintaining the upsampling ratio as an integer - 2x, 4x, 8x, 16x, etc...

Thank you for considering this feature!

**Comment by yo3fxy** [comment]
@GitHubGeek to treat upsampling factor as a base rate multiplier, change line 118 in file setrate_util.c to

`samplerate = (rate % 48000 ? 44100 : 48000) * upsampling_factor;`

@marcoevang my intended use for this feature was always to treat upsampling factor as a base rate multiplier.

**Comment by marcoevang** [comment]
@GitHubGeek 
> Is it possible to make the factor dynamic, similar to how Squeezelite resample work?

I investigated on this. Unfortunately, the highest sample rate supported by the DAC cannot be reliably read in a general way, as it depends on the specific device and its driver. Threrefore I gave up implementing this feature.
However, if you know the highest sample rate supported by your DAC, you can easily achieve your goal by implementing the mod suggested by @GitHubGeek.

**Comment by marcoevang** [comment]
@GitHubGeek 
> samplerate = (rate % 48000 ? 44100 : 48000) * upsampling_factor;

This is an interesting mod that I will implement as soon as I have some spare time.
I have reopened this issue.
